pkg/metrics/metrics.go

```diff
@@ -0,0 +1,218 @@
+			},
+			[]string{"remote_address"},
+		)
+
+		oidcAuthCount = prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: promNamespace,
+				Name:      "oidc_authentication_count",
+				Help:      "The count for OIDC authentication. Authenticated requests are 1, else 0.",
+			},
+			[]string{"authenticated", "remote_address", "user"},
+		)
+
+		tokenReviewDuration = prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Namespace: promNamespace,
+				Name:      "token_review_duration_seconds",
+				Help:      "The duration in seconds for a token review lookup. Authenticated requests are 1, else 0.",
+				Buckets:   prometheus.LinearBuckets(.000, .05, 30),
+			},
+			[]string{"authenticated", "code", "remote_address", "user"},
+		)
+	)
+
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(clientRequests)
+	registry.MustRegister(clientDuration)
+	registry.MustRegister(serverRequests)
+	registry.MustRegister(serverDuration)
+	registry.MustRegister(oidcAuthCount)
+	registry.MustRegister(tokenReviewDuration)
+
+	return &Metrics{
+		registry: registry,
+
+		clientRequests: clientRequests,
+		clientDuration: clientDuration,
+
+		serverRequests: serverRequests,
+		serverDuration: serverDuration,
+
+		oidcAuthCount:       oidcAuthCount,
+		tokenReviewDuration: tokenReviewDuration,
+	}
+}
+
+// Start will register the Prometheus metrics, and start the Prometheus server
+func (m *Metrics) Start(listenAddress string) error {
+	router := http.NewServeMux()
+	router.Handle("/metrics", promhttp.HandlerFor(m.registry, promhttp.HandlerOpts{}))
+
+	ln, err := net.Listen("tcp", listenAddress)
+	if err != nil {
+		return err
+	}
+
+	m.Server = &http.Server{
+		Addr:           ln.Addr().String(),
+		ReadTimeout:    8 * time.Second,
+		WriteTimeout:   8 * time.Second,
+		MaxHeaderBytes: 1 << 15, // 1 MiB
+		Handler:        router,
+	}
+
+	go func() {
+		klog.Infof("serving metrics on %s/metrics", ln.Addr())
+
+		if err := m.Serve(ln); err != nil {
+			klog.Errorf("failed to serve prometheus metrics: %s", err)
+			return
+		}
+	}()
+
+	return nil
+}
+
+func (m *Metrics) Shutdown() error {
+	// If metrics server is not started than exit early
+	if m.Server == nil {
+		return nil
+	}
+
+	klog.Info("shutting down Prometheus metrics server...")
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	if err := m.Server.Shutdown(ctx); err != nil {
+		return fmt.Errorf("prometheus metrics server shutdown failed: %s", err)
+	}
+
+	klog.Info("prometheus metrics server gracefully stopped")
+
+	return nil
+}
+
+func (m *Metrics) ObserveClient(code int, path, remoteAddress string, duration time.Duration) {
+	m.clientRequests.With(prometheus.Labels{
+		"code":           strconv.Itoa(code),
+		"path":           path,
+		"remote_address": remoteAddress,
+	}).Inc()
+
+	m.clientDuration.With(prometheus.Labels{
+		"remote_address": remoteAddress,
+	}).Observe(duration.Seconds())
```

Comment:
I would still include `code` here, and maybe `path` too. It can be very useful to see that it is a particular response code that is causing slow responses. Same for the path, but maybe you are concerned about the label size?

Reply:
Yes, with the duration included this makes the number of dimensions really high.

Reply:
I would still include `code`, but I don't actually know what the specific recommendations on sizes are.